### PR TITLE
many: check installation of slots and plugs against declarations

### PR DESCRIPTION
**Pull request by pedronis**
Also this fill out more the slots side of base-declaration with *-installation rules.

In general see the TODO at the bottom of it in basedeclaration.go with what is left TODO.

### CODE REVIEW

File: interfaces/builtin/basedeclaration.go [hunk 1]
```diff
@@ -34,100 +34,258 @@ series: 16
...
+      slot-snap-type:
+        - app
+        - gadget
   cups-control:
+    allow-installation:
+      slot-snap-type:
+        - core
     deny-auto-connection: true
   docker:
+    allow-installation: false
```
Comment (jdstrand) [inline interfaces/builtin/basedeclaration.go]: Without snap declarations, this is going to make the docker snap uninstallable. I think this needs to be this for now:

```
docker:
  deny-auto-connection: true
```

File: interfaces/builtin/basedeclaration.go [hunk 2]
```diff
@@ -34,100 +34,258 @@ series: 16
...
+        - core
+  upower-observe:
+    allow-installation:
+      slot-snap-type:
+        - core
+  x11:
+    allow-installation:
+      slot-snap-type:
+        - core
 `
```
Comment (jdstrand) [inline interfaces/builtin/basedeclaration.go]: This is all correct (except docker, see above) for while we don't have snap declarations.

File: overlord/ifacestate/helpers.go [hunk 3]
```diff
@@ -295,3 +295,33 @@ func getConns(st *state.State) (map[string]connState, error) {
 func setConns(st *state.State, conns map[string]connState) {
 	st.Set("conns", conns)
 }
+
+// CheckInterfaces checks whether plugs and slots of snap are allowed for installation.
+func CheckInterfaces(st *state.State, snap *snap.Info) error {
```
Comment (jdstrand) [inline overlord/ifacestate/helpers.go]: CheckInterfaces() is a pretty generic name. I wonder if CheckInstallationInterfaces() or something else would be better.

File: overlord/snapstate/check_snap.go [hunk 4]
```diff
@@ -79,17 +79,25 @@ func checkSnap(st *state.State, snapFilePath string, curInfo *snap.Info, flags F
...
+	if CheckInterfaces != nil {
+		if err := CheckInterfaces(st, s); err != nil {
+			return err
+		}
+	}
+
 	switch s.Type {
 	case snap.Type(""), snap.TypeApp, snap.TypeKernel:
 		// "" used in a lot of tests :-/
+
```
Comment (pedronis) [inline overlord/snapstate/check_snap.go]: spotted this extra whitespace, fixing

**Pull request merged**